internal/provider/cluster_resource.go

```diff
@@ -716,23 +701,65 @@ func (r *clusterResource) Update(
 			return
 		}
 
-		upgradeStatus := client.CLUSTERUPGRADESTATUSTYPE_MAJOR_UPGRADE_RUNNING
-		cmp := compareCrdbVersions(stateVersion, planVersion, &resp.Diagnostics)
-		if cmp < 0 {
-			// Make sure we're rolling back to the previous version.
-			if cmp < -1 {
-				resp.Diagnostics.AddError("Invalid rollback version", "Can only roll back to the previous version.")
+		// Next, if we are upgrading, it must be a valid upgrade from the
+		// current version, according to the ListMajorClusterVersions API.
+		if upgrading, err := isUpgrade(stateVersion, planVersion); upgrading {
+			var currentVersionInfo client.ClusterMajorVersion
+			for _, v := range apiResp.Versions {
+				if v.Version == stateVersion {
+					currentVersionInfo = v
+					break
+				}
+			}
+
+			var validUpgrade bool
+			for _, v := range currentVersionInfo.AllowedUpgrades {
+				if v == planVersion {
+					validUpgrade = true
+					break
+				}
+			}
+			if !validUpgrade {
+				resp.Diagnostics.AddError("Invalid CockroachDB version",
+					fmt.Sprintf(
+						"Cannot change major version to '%s'. Valid upgrade versions include [%s]",
+						planVersion,
+						strings.Join(currentVersionInfo.AllowedUpgrades, "|")))
 				return
 			}
-			upgradeStatus = client.CLUSTERUPGRADESTATUSTYPE_ROLLBACK_RUNNING
-		} else if cmp > 1 {
-			resp.Diagnostics.AddError("Invalid upgrade version", "Can't skip versions. Upgrades must be performed one version at a time.")
+		} else if err != nil {
+			resp.Diagnostics.AddError("Invalid CockroachDB version", err.Error())
 			return
 		}
 
+		if downgrade, _ := isDowngrade(stateVersion, planVersion); downgrade {
+			var targetVersionInfo client.ClusterMajorVersion
+			for _, v := range apiResp.Versions {
+				if v.Version == planVersion {
+					targetVersionInfo = v
+					break
+				}
+			}
+
+			var validDowngrade bool
+			for _, v := range targetVersionInfo.AllowedUpgrades {
+				if v == stateVersion {
+					validDowngrade = true
+					break
+				}
+			}
+			if !validDowngrade {
+				resp.Diagnostics.AddError("Invalid CockroachDB version",
+					fmt.Sprintf(
+						"Cannot roll back to to '%s'.",
+						planVersion))
+				return
```

Comment:
I was going to ask if we can list the valid rollback version here but... do we actually not know it at this point?  If we let this pass through, it seems like we can give a better error.  WDYT?

Reply:
We don't include the rollback version in the CC API clusters response. We could, there's probably no harm in doing so. Do you think that's valuable to add?

Reply:
Without all this error handling, we get:

```
cockroach_cluster.example: Modifying... [id=2dc9a589-1f20-412c-952b-522d5184a3f3]
cockroach_cluster.example: Still modifying... [id=2dc9a589-1f20-412c-952b-522d5184a3f3, 10s elapsed]
cockroach_cluster.example: Still modifying... [id=2dc9a589-1f20-412c-952b-522d5184a3f3, 20s elapsed]
╷
│ Error: Error updating cluster
│
│   with cockroach_cluster.example,
│   on main.tf line 49, in resource "cockroach_cluster" "example":
│   49: resource "cockroach_cluster" "example" {
│
│ cluster can only be rolled back to the previous version, expected: v24.1
```

I don't know why this took 20+ seconds to get this? But it seems like a reasonable error, so I'll remove the checks in the tf provider

Reply:
Here is the request: https://us5.datadoghq.com/apm/trace/8538d437c1868d2218b1a60d611750c1?graphType=flamegraph&shouldShowLegend=true&sort=time&spanID=1655416649021547938&timeHint=1728402657004.241


350ms for the actual api request.    I think it probably took 20s because of all the other requests that happen during a tf run.  It looks like there are about 20 requests to the same cluster endpoint.  API caching via terraform isn't a good story currently.

<img width="1502" alt="Screenshot 2024-10-09 at 8 41 53 AM" src="https://github.com/user-attachments/assets/225a82a7-9793-47d6-be1d-0676118e0048">